examples/coarse_alignment.py

```diff
@@ -86,6 +89,46 @@
     current_shift += shift
     coarse_shifts[i + 1] = current_shift
 
+# optimize tilt axis angle
+grid_resolution = 1
+tilt_axis_grid = CubicBSplineGrid1d(resolution=grid_resolution, n_channels=1)
+tilt_axis_grid.data = torch.tensor([TILT_AXIS_ANGLE_PRIOR, ] * grid_resolution, dtype=torch.float32)
+interpolation_points = torch.linspace(0, 1, len(tilt_series))
+
+common_lines_optimiser = torch.optim.Adam(
+    tilt_axis_grid.parameters(),
+    lr=1,
+)
+
+for epoch in range(100):
+    # interpolate the grid
+    tilt_axis_angles = tilt_axis_grid(interpolation_points)
+
+    # for common lines each 2d image is projected perpendicular to the tilt axis, thus add 90 degrees
+    R = Rz(tilt_axis_angles + 90, zyx=False)[:, :2, :2]
+
+    projections = []
+    for i in range(len(tilt_series)):
+        p = project_real_2d(tilt_series[i] * coarse_alignment_mask, R[i:i+1])
+        projections.append((p - p.mean()) / p.std())
+
+    common_lines_optimiser.zero_grad()
+    loss = 0
+    for x, y in combinations(projections, 2):
+        loss = loss - (x * y).sum() / y.numel()
```

Comment:
I'm wondering if this is the right form for the loss function, I see how it should behave and it seems to optimise which is good! Typically we would minimise the sum of squared differences between all projections, is this way better? if so, how come? :-)

Given that the projections are small maybe something like this, starting from projections with shape `b, d`

```python
squared_differences = (projections - einops.rearrange(projections, 'b d -> b 1 d')) ** 2
loss = einops.reduce(squared_differences, 'b1 b2 d -> 1', reduction='sum')
```

Some duplicate work but the vectorisation should overall speed it up 🤷🏼‍♂️

Reply:
Good point, that is much nicer :) Updated now